teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md

```diff
@@ -0,0 +1,483 @@
+ 1. Use Git commits as sources, and generates in-toto provenance for a TaskRun that builds an image.
+   
+    A Task example **without** structured results and params support can be like:
+    ``` yaml
+    apiVersion: tekton.dev/v1beta1
+    kind: Task
+    spec:
+      params:
+        - name: CHAINS-GIT_COMMIT
+          type: string
+          description: git commit sha
+        - name: CHAINS-GIT_URL
+          type: string
+          description: git commit url
+        - name: FOO_IMAGE
+          type: string
+          description: artifact image
+      ...
+      results:
+        - name: IMAGE_DIGEST
+          description: digest of image
+        - name: IMAGE_URL
+          description: url of image
+    ```
+    A Task example `with` structured results support will be like:
+    ``` yaml
+    apiVersion: tekton.dev/v1beta1
+    kind: Task
+    spec:
+      ...
+      results:
+        - name: ARTIFACT-INPUTS_git-vcs
+          type: object
+          description: |
+            The source distribution
+            * uri: resource uri of the artifact.
+            * digest: revision digest in form of algorithm:digest.
+          properties:
+            uri:
+              type: string
+            digest:
+              type: string
+        - name: ARTIFACT-OUTPUTS_oci_image
+          type: object
+          description: N/A
+          properties:
+            uri:
+              type: string
+            digest:
+              type: string
+    ```
+    And its corresponding TaskRun example will be like:
+    ``` yaml
+    TaskRun:
+      ...
+      results:
+        - name: ARTIFACT-INPUTS_git-vcs
+          value:
+            uri: git+https://github.com/foo/bar.git
+            digest: sha256:abc
+        - name: ARTIFACT-OUTPUTS_oci-image
+          value:
+            uri: gcr.io/somerepo/someimage
+            digest: sha512:abc
+    ```
+The generated intoto provenance will contain the following:
+```
+subjects: [{"name": "gcr.io/somerepo/someimage", "digest": {"sha256": "abc"}}]
+...
+materials: [{
+  "uri": "git+https://github.com/foo/bar.git",
+  "digest": {"sha1": "abc..."}
+}]
+```
+
+2. Use Perforce and images as sources, and generates in-toto provenance for a TaskRun that builds a Maven package.
+``` yaml
+results:
+  - name: ARTIFACT-INPUTS_perforce-vcs
+    value:
+      uri: http://myp4web:8080/depot/main/atlas/
+      digest: sha256:abc
+  - name: ARTIFACT-OUTPUTS_maven-pkg
+    value:
+      uri: us-west4-maven.pkg.dev/test-project/test-repo/com/google/guava/guava/31.0/guava-31.0.jar
+      digest: sha256:abc
+  - name: ARTIFACT-OUTPUTS_maven-pom
+    value:
+      uri: us-west4-maven.pkg.dev/test-project/test-repo/com/google/guava/guava/31.0/guava-31.0.pom
+      digest: sha256:def
+  - name: ARTIFACT-OUTPUTS_maven-src-pkg
+    value:
+      uri: us-west4-maven.pkg.dev/test-project/test-repo/com/google/guava/guava/31.0/guava-31.0-sources.jar
+      digest: sha256:xyz
```

Comment:
Both the algorithm name and the digest would not have ",". Could we have a comma separated list of sha:digest which we could parse out. This is just to be feature complete wrt to the DigestSet in the SLSA specification.

Reply:
That's a good point. I can definitely add it in the implementation, but let me bring this up for discussion. I will first implement the basic feature